### DISCUSSION TITLE
Tracker: LINQ query expression proposals

### DISCUSSION THREAD

**Discussion by orthoxerox** [dotnet/csharplang#116] [General]
Keyword (Method) | Issue | Stage
-- | - | -
`(take) while` (`TakeWhile()`, `SkipWhile()`) | https://github.com/dotnet/csharplang/issues/94, #592 | Discussion
`do` (`foreach`) | https://github.com/dotnet/csharplang/issues/101 | Championed
(`Aggregate()`) | #333 | 
(`All(), Any()`) | #333 | 
(`Average()`, `Sum()`, `Count()`, `Min()`, `Max()`) | https://github.com/dotnet/csharplang/issues/333 | 
(`Concat()`, `Except()`, `Intersect()`, `Union()`) |  | 
(`Distinct()`) |  | 
(`First()`, `FirstOrDefault()`, `Single()`, `SingleOrDefault()`, `Last()`, `LastOrDefault()`) | https://github.com/dotnet/csharplang/issues/119 and also https://github.com/dotnet/csharplang/issues/333 | Discussion
`take`, `skip` (`Take()`, `Skip()`) | #592 | Discussion
`with` (`Zip()`) | https://github.com/dotnet/csharplang/issues/117 | Discussion
`apply` to call `ToList()` etc | https://github.com/dotnet/csharplang/issues/333 | Discussion
`leftjoin, rightjoin` | https://github.com/dotnet/csharplang/issues/361 | Discussion

**Comment by alrz** [comment]
or [#100 (comment)](https://github.com/dotnet/roslyn/issues/100/#issuecomment-239918884)

**Comment by phi1010** [comment]
For Linq-Methods using two IEnumerables (such as Zip, Union, Except, Concat, Intersect) I'd rather see an extendable syntax allowing to use userdefined extention methods, as suggested in https://github.com/dotnet/csharplang/issues/117#issuecomment-285788455

**Comment by MgSam** [comment]
And what happens for the dozen extension methods I've defined myself on `IEnumerable`?

**Comment by alrz** [comment]
I'd rather see a spec for [#100](https://github.com/dotnet/roslyn/issues/100/#issuecomment-239918884) as it can cover most (all?) of these. (except for #101 which is just a statement form of query expressions and doesn't really belong here).

**Comment by orthoxerox** [comment]
@alrz why not write one?

**Comment by alrz** [comment]
@orthoxerox 

This has some interaction with #101 so I think it's better to wait for a proposal for #101 first.

**Comment by phi1010** [comment]
@alrz: The idea is definitly not a bad one, but I miss the possibility to continue with my query after executing the do-operation. It doesn't make much sense on a call to Single/First/Last/..., but when I use `do WeirdPseudoSort()`, I'd like to be still able to access the variables I defined above by using let or select clauses, without writing this:

```cs
from x in (
    from x1 in something
    let x2 = dosomething(x1)
    select new{x1, x2}
    do WeirdPseudoSort() // still returns an IEnumberable<T>, not a single T
)
let x1 = x.x1;
let x2 = x.x2;
let x3 = dosomethingelse(x1,x2)
select x3;
```

I'd like to see a keyword that allows packing the current contect (all variable values from inside the linq statement) into a tuple then processing those tuples, unpacking them before evaluating the next linq query.

**Comment by alrz** [comment]
@phi1010 (also @orthoxerox)

> I miss the possibility to continue with my query after executing 

I think `do` should be able to continue, e.g.
```cs
var q =
  from x1 in something
  let x2 = func1(x1)
  do TakeWhile(!x1.foo)
  select func2(x1, x2);

var q = something
  .Select(x1 => (x1, x2: func1(x1)))
  .TakeWhile(t => !t.x1.foo)
  .Select(t => func2(t.x1, t.x2));
```
That makes it a more complicated because currently query terminations (group and select) need an `into` clause to be continued. Though it would be interesting to see how this can fit into the query syntax.

**Comment by orthoxerox** [comment]
@alrz see the discussion in #333, it is able to continue the expression, but Matt considers the can of worms it opens to be too big to approve of this.
